File: c170-179/c179/main.py
import math

def union(l1, l2):
    return [l1[i] or l2[i] for i in range(len(l1 + l2) // 2)]

for i in range(int(input())):
    n, m = list(map(int, input().split()))

    l = list(map(int, input().split()))
    n_happy, happy_boys = l[0], l[1:]
    boys = [i in happy_boys for i in range(n)]

    l = list(map(int, input().split()))
    m_happy, happy_girls = l[0], l[1:]
    girls = [i in happy_girls for i in range(m)]

    if n_happy == 0 and m_happy == 0:
        print("NO")
        continue

    if (gcd := math.gcd(n, m)) == 1 and (m_happy or n_happy):
        print("YES")
        continue

    step = [0] * gcd

    for i in range(0, m // gcd):
        step = union(step, girls[i * gcd: (i + 1) * gcd])
    for i in range(0, n // gcd):
        step = union(step, boys[i * gcd: (i + 1) * gcd])
    
    if all(step):
        print("YES")
    else:
        print("NO")

File: c170-179/c179/main.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

unsigned short int find_gcd(unsigned short int n, unsigned short int m) {
    unsigned short int r;
    while (n > 0) {
        r = m % n;
        m = n;
        n = r;
    }

    return m;
}

unsigned short int all(bool* L, unsigned short int N) {
    bool a = true;
    for (unsigned short int i = 0; i < N && a; i++) {
        a = a && L[i];
    }
    return a;
}

int main() {
    int T;
    scanf("%d", &T);

    unsigned short int n, m, n_happy, m_happy, gcd, dummy;
    bool step[100];

    while (T--) {
        scanf("%hu %hu", &n, &m);
        
        gcd = find_gcd(n, m);
        memset(step, false, gcd * sizeof(bool));
        
        scanf("%hu", &n_happy);
        while (n_happy--) {
            scanf("%hu", &dummy);
            step[dummy % gcd] = true;
        }

        scanf("%hu", &m_happy);
        while (m_happy--) {
            scanf("%hu", &dummy);
            step[dummy % gcd] = true;
        }

        if (all(step, gcd)) {
            puts("YES");
        } else {
            puts("NO");
        }
    }
}
